pkg/controller/node/nodecontroller.go

```diff
@@ -503,19 +526,49 @@ func (nc *NodeController) monitorNodeStatus() error {
 			// Check eviction timeout against decisionTimestamp
 			if observedReadyCondition.Status == api.ConditionFalse &&
 				decisionTimestamp.After(nc.nodeStatusMap[node.Name].readyTransitionTimestamp.Add(nc.podEvictionTimeout)) {
-				if nc.evictPods(node) {
-					glog.V(2).Infof("Evicting pods on node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout)
+				nodeNotReadyTaint := api.Taint{
+					Key:       unversioned.TaintNodeNotReady,
+					Effect:    api.TaintEffectNoExecute,
+					AddedTime: nc.now(),
+				}
+				added, err := tryAddTaintToNode(nc.kubeClient, node.Name, nodeNotReadyTaint)
+				if err != nil {
+					glog.Errorf("Failed to try add taint %s to node %s: %v", nodeNotReadyTaint.ToString(), node.Name, err)
+				}
+				if added {
+					glog.V(4).Infof("Added nodeNotReady taint to node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout)
 				}
 			}
 			if observedReadyCondition.Status == api.ConditionUnknown &&
 				decisionTimestamp.After(nc.nodeStatusMap[node.Name].probeTimestamp.Add(nc.podEvictionTimeout)) {
-				if nc.evictPods(node) {
-					glog.V(2).Infof("Evicting pods on node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout-gracePeriod)
+				nodeUnreachableTaint := api.Taint{
+					Key:       unversioned.TaintNodeUnreachable,
+					Effect:    api.TaintEffectNoExecute,
+					AddedTime: nc.now(),
+				}
+				added, err := tryAddTaintToNode(nc.kubeClient, node.Name, nodeUnreachableTaint)
+				if err != nil {
+					glog.Errorf("Failed to try add taint %s to node %s: %v", nodeUnreachableTaint.ToString(), node.Name, err)
+				}
+				if added {
+					glog.V(4).Infof("Added unreachable taint to node %s: %v is later than %v + %v", node.Name, decisionTimestamp, nc.nodeStatusMap[node.Name].readyTransitionTimestamp, nc.podEvictionTimeout-gracePeriod)
 				}
 			}
 			if observedReadyCondition.Status == api.ConditionTrue {
-				if nc.cancelPodEviction(node) {
-					glog.V(2).Infof("Node %s is ready again, cancelled pod eviction", node.Name)
+				nodeNotReadyTaint := api.Taint{
+					Key:    unversioned.TaintNodeNotReady,
+					Effect: api.TaintEffectNoExecute,
+				}
+				nodeUnreachableTaint := api.Taint{
+					Key:    unversioned.TaintNodeUnreachable,
+					Effect: api.TaintEffectNoExecute,
+				}
+				removed, err := tryRemoveTaintsOffNode(nc.kubeClient, node.Name, nodeNotReadyTaint, nodeUnreachableTaint)
```

Comment:
I'm not a fan of sending all those updates separately. Can all this logic go into 'tryUpdateNodeStatus' function, when we actually do all the object updates. This way we can compress all those API requests into a single one.